### PR TITLE
Added base structure for multiple currencies

### PR DESCRIPTION
**Pull request by kazazor**
# What this PR is all about?

* Added target currency switcher (added 7 different options)
* Some things to know:
  * The currencies request always comes BEFORE the coins request. No special handling in case it doesn't
  * I did not take care of a situation we have a failure in the fetch currencies API. In this case I would actually want to just dispatch the action again, need to understand what to do there..
  * I do not save crypto target currencies in localstorage, this is to make things simpler in an already big PR. Opened a ticket for that: https://github.com/kazazor/coinsmarket/issues/105
* I did everything in the coins reducer / service even that the currencies are a little bit different. The reason is that I needed the state this reducer is in charge of for the coins data to be updated upon switch of the target currency. If you have a better idea be my guest :)

### Additional information for the reviewer

A man with severe OCD and a phobia of germs attempted to commit suicide at the age of 19 via a gunshot wound to the head. Instead of killing him, the bullet eliminated his mental illness without any other brain damage.

### Issue
Github issue: #43

### CODE REVIEW

**Comment by kazazor** [comment]
@OmriAharon - Enjoy the review :)
I feel like I wouldn't want to merge this one without a review..

File: src/components/TargetCurrencyMenu.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    this.state = {
+      anchorEl: undefined,
+      open: false,
+      targetCurrencyCode: props.targetCurrency.code
+    };
+  }
+
+  _handleClick(event) {
+    this.setState({ open: true, anchorEl: event.currentTarget});
```
Comment (OmriAharon) [inline src/components/TargetCurrencyMenu.js]: Don't we want to toggle the open state here?

File: src/components/TargetCurrencyMenu.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    this.setState({ open: false });
+  }
+
+  _renderMenuItems() {
+    return Object.values(targetCurrencies).map((currency, index) => {
+      return (
+        <MenuItem
+          key={currency.code}
+          selected={currency.code === this.state.targetCurrencyCode}
+          onClick={(event) => this._handleMenuItemClick.call(this, event, currency)}>
```
Comment (OmriAharon) [inline src/components/TargetCurrencyMenu.js]: Just curious why not `(event) => this._handleMenuItemClick(event, currency)` ?

File: src/helpers/targetCurrencies.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  [rub.code]: rub
+};
+
+const targetCurrencies = Object.assign({
+  [btc.code]: btc,
+  [eth.code]: eth
+}, regularTargetCurrencies);
+
+export const getRegularTargetCurrency = (code) => {
+  return code ? regularTargetCurrencies[code] : undefined;
```
Comment (OmriAharon) [inline src/helpers/targetCurrencies.js]: NIT -> `return code && regularTargetCurrencies[code];`

File: src/redux/actions/coinsApiActions.js [hunk 4]
```diff
@@ -30,3 +30,29 @@ export const fetchCoinsList = () => ({
     ]
   }
 });
+
+export const fetchRegularCurrencies = (currenciesToFetch) => {
+  const currencies = Object.values(currenciesToFetch).reduce((output, currency) => {
+    return `${output}${currency.code},`;
+  }, '');
```
Comment (OmriAharon) [inline src/redux/actions/coinsApiActions.js]: Won't `Object.keys(currenciesToFetch).join(',')` be easier?

File: src/redux/actions/coinsApiActions.js [hunk 5]
```diff
@@ -30,3 +30,29 @@ export const fetchCoinsList = () => ({
...
+  return ({
+    [CALL_API]: {
+      endpoint: `${config.ORIGINS.FIXER_IO}/latest?base=USD&symbols=${currencies}`,
+      method: 'GET',
+      types: [
+        Actions.FETCH_REGULAR_CURRENCIES,
+        {
+          type: Actions.FETCH_REGULAR_CURRENCIES_SUCCESS
+        },
+        Actions.FETCH_REGULAR_CURRENCIES_FAILURE
```
Comment (OmriAharon) [inline src/redux/actions/coinsApiActions.js]: I think you can do `[Actions.FETCH_REGULAR_CURRENCIES, Actions.FETCH_REGULAR_CURRENCIES_SUCCESS, Actions.FETCH_REGULAR_CURRENCIES_FAILURE]` here. (without the `{ type: ... }`)

File: src/redux/reducers/coinsReducer.js [hunk 6]
```diff
@@ -89,13 +122,25 @@ const CoinsReducer = (state = initialState, action) => {
...
 
     for (index = 0, rankIndex = coins.length; index < unSortedCoins.length; index++, rankIndex++) {
-      coins.push(ValuePair.parse(unSortedCoins[index], action.meta.locale, rankIndex + 1));
+      const pair = ValuePair.parse(unSortedCoins[index], action.meta.locale, rankIndex + 1, state.targetCurrency);
+      coins.push(pair);
+
+      // If the current coins is also a possible target currency, update its USD rate
+      if (targetCurrencies[pair.baseCurrency.code]) {
+        targetCurrencies[pair.baseCurrency.code].factorFromUSD = 1 / pair.price;
+      }
```
Comment (OmriAharon) [inline src/redux/reducers/coinsReducer.js]: Code duplication with lines 125-132

File: src/components/TargetCurrencyMenu.js (hunk 1 above)
Comment (kazazor) [inline src/components/TargetCurrencyMenu.js]: I am doing it.. Setting the `open` state to `true`

This is the button that opens the menu

File: src/components/TargetCurrencyMenu.js (hunk 2 above)
Comment (kazazor) [inline src/components/TargetCurrencyMenu.js]: Well I need the `this` to be the component and not the node element.. So I can add `bind` in the `constructor` instead if it helps

File: src/redux/actions/coinsApiActions.js (hunk 4 above)
Comment (kazazor) [inline src/redux/actions/coinsApiActions.js]: It is :)

File: src/redux/reducers/coinsReducer.js (hunk 6 above)
Comment (kazazor) [inline src/redux/reducers/coinsReducer.js]: Yeah I knew about it and was too lazy.. you got me :)

**Comment by kazazor** [comment]
@OmriAharon , I fixed your comments. But I would really love to hear what do you think about:

* The currencies request always comes BEFORE the coins request. No special handling in case it doesn't
* I did not take care of a situation we have a failure in the fetch currencies API. In this case I would actually want to just dispatch the action again, need to understand what to do there..

**Comment by kazazor** [comment]
I'll merge it now anyway, but still wants to hear your thoughts :)

**Pull request merged**